Building quality apps/Unit tests.md

```diff
@@ -0,0 +1,155 @@
+
+![testing_pyramid.png](../img/testing_pyramid.png)
+
+The testing pyramid above consists of three main parts:
+
+* Unit tests
+* Integration tests
+* UI Tests
+
+As one can see the foundation of the pyramid are Unit tests which are the highest in number in most projects and in this chapter we will concentrate only on Unit tests. 
+
+## Unit tests
+
+Unit tests are tests that are isolated, fast in execution and cheap to write. They run on a plain JVM and because of that it can't have any dependencies that are not available out of-the-box on the JVM. In order to achieve this we have to design our codebase in a way that allows us to neatly separate units that we want to test. Sometimes it is very hard to achieve this but luckily we have some tools in our arsenal that make our life easier, like jUnit and Mockito.
+
+## Anatomy of a unit test
+
+In order to write better tests that are easy to read and give us the exact amount of information that we need to actually understand the test it is paramount to choose an appropriate naming and structure for our tests. Let's take a look at an example test which is has no naming convention nor a defined structure:
+
+```kotlin
+@Test
+fun `login`() {
+    val username = "Ivica"
+    val password = "123456"
+    val expectedUser = User("Ivica", Status.REGISTERED)
+    whenever(loginInteractor.login(username, password)).thenReturn(LoginResult(expectedUser, "Token"))
+    val actualUser = loginUseCase.login(username, password)
+    assertEquals(expectedUser, actualUser)
+}
+```
+The above test does not give us any information about what this code is supposed to test and in which scenario it is going to run. If we look at the test structure it is not so easy to understand what is happening here and at first glance it seems messy. Keep in mind that this test is a simple test, so in more complex examples this could get even messier. 
+
+### Test method naming
+
+The first thing that is wrong in the above test is the test method name. We want to have a name which give us information about the scenario which we are testing, the end result that is expected and what method are we testing. **There are a lot of well defined naming conventions out there in the wild and each one of them has some pros and cons so the most important thing is to pick one that makes most sense to use, but after you pick one try to be consistent across the codebase!**
+
+For the purpose of this example we will use a naming convention that follows this pattern `methodName ... should ... when`. Now let's try to improve the naming of the test in a way that it gives us more information about the scenario that is being tested here. 
+
+```kotlin
+@Test
+fun `login should return a user object when the login is successful`() { ... }
```

Comment:
I understand your point and it is totally fine with me if we decide to do this! But I have mixed feelings about this, because the beginning of the sentence is the method name which we test and therefore it seems more accurate writing the first letter in lower case style. Also, another thing I am worried about here is that people are used to write method names in lower case so the above approach may not come natural to them, but maybe it is a long shot.

I am curious what others think about this?

Reply:
You raise a good point. Since this was just a one-word method name, I forgot that the first word is the method name. I agree with your standpoint! 👍